File: apps/web/src/lib/money-conversion.test.ts
import { describe, expect, it } from 'vitest';

import {
  convertMinorBetweenCurrenciesSV,
  fromBsMinorSV,
  pickExchangeRateForDateSV,
  toBsMinorSV,
} from './money-conversion';

describe('money-conversion', () => {
  it('should convert USD minor to BS minor using rate 50 Bs/USD', () => {
    expect(toBsMinorSV(3000, 'USD', 50)).toBe(150_000);
  });

  it('should convert BS minor to USD minor using rate 50 Bs/USD', () => {
    expect(fromBsMinorSV(150_000, 'USD', 50)).toBe(3000);
  });

  it('should convert USD obligation to BS settlement', () => {
    expect(
      convertMinorBetweenCurrenciesSV(3000, 'USD', 'BS', 50, 1),
    ).toBe(150_000);
  });

  it('should pick rate for reservation calendar date', () => {
    const RATES = [
      {
        currency: 'USD',
        rateToBs: 48,
        effectiveDate: '2026-05-16T00:00:00.000Z',
      },
      {
        currency: 'USD',
        rateToBs: 50,
        effectiveDate: '2026-05-17T00:00:00.000Z',
      },
    ];
    const PICKED = pickExchangeRateForDateSV(RATES, 'USD', '2026-05-17');
    expect(PICKED?.rateToBs).toBe(50);
  });
});

File: apps/mobile/test/core/formatting/money_conversion_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:cuadrala_mobile/src/core/formatting/money_conversion.dart';
import 'package:cuadrala_mobile/src/core/models/currency_code.dart';

void main() {
  group('convertMinorBetweenCurrenciesSV', () {
    test('should convert USD to BS using rates', () {
      expect(
        convertMinorBetweenCurrenciesSV(
          10000,
          CurrencyCode.usd,
          CurrencyCode.bs,
          50,
          1,
        ),
        500000,
      );
    });

    test('should return same amount when currencies match', () {
      expect(
        convertMinorBetweenCurrenciesSV(
          1500,
          CurrencyCode.usd,
          CurrencyCode.usd,
          50,
          1,
        ),
        1500,
      );
    });
  });

  group('pickExchangeRateForDateSV', () {
    test('should pick exact date rate', () {
      final picked = pickExchangeRateForDateSV(
        const [
          ExchangeRateRow(
            currency: 'USD',
            rateToBs: 40,
            effectiveDate: '2026-05-10',
          ),
          ExchangeRateRow(
            currency: 'USD',
            rateToBs: 50,
            effectiveDate: '2026-05-17',
          ),
        ],
        CurrencyCode.usd,
        '2026-05-17',
      );
      expect(picked?.rateToBs, 50);
    });
  });
}
